cupyx/scipy/special/_logsumexp.py

```diff
@@ -0,0 +1,74 @@
+import cupy as cp
+
+
+__all__ = ["log_softmax"]
+
+
+_log_preamble = '''
+
+template <typename T>
+__device__ T _log(T a)
+{
+    return log(a);
+}
+
+'''
+
+
+_preamble = _log_preamble+'''
+
+template <typename T>
+__device__ T _post_map(T a, T* y)
+{
+    *y = _log(static_cast<float>(a));
+    return *y;
+}
+
+'''
+
+
+_log_softmax_kernel = cp._core.ReductionKernel(
+   'T x1',
+   'T y',
+   'exp(static_cast<float>(x1))',
+   'a + b',
+   'y = _post_map(a, &y)',
```

Comment:
Let's use `exp` and `log` without casting to float to make things simple at first.

Reply:
```py
_log_softmax_kernel = cp._core.ReductionKernel(
   'T x1',
   'T y',
   'exp(x1)',
   'a + b',
   'y = log(a)',
   '0',
   name='log_softmax',
   preamble=_preamble
)
```

Reply:
This was my first try. But I got the compilation error saying:
<details>
<summary>Error Log (Click Please)</summary>

<p>

```python
__________________________ TestLogSoftmax.test_log_softmax_large_inputs ____________________________
 
self = <cupyx_tests.scipy_tests.special_tests.test_log_softmax.TestLogSoftmax object at 0x7f36ad3341c0>
xp = <module 'cupy' from '/home/khushi/Documents/opensource/cupy/cupy/__init__.py'>
scp = <module 'cupyx.scipy' from '/home/khushi/Documents/opensource/cupy/cupyx/scipy/__init__.py'>
dtype = <class 'numpy.int8'>
 
    @testing.for_all_dtypes()
    @testing.numpy_cupy_allclose(scipy_name='scp', atol=atol, rtol=rtol)
    def test_log_softmax_large_inputs(self, xp, scp, dtype):
        a = xp.arange(4, dtype=dtype)
>       return scp.special.log_softmax(a)
 
tests/cupyx_tests/scipy_tests/special_tests/test_log_softmax.py:23: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
cupyx/scipy/special/_logsumexp.py:71: in log_softmax
    out = _log_softmax_kernel(tmp, axis=axis, keepdims=True)
cupy/_core/_reduction.pyx:752: in cupy._core._reduction.ReductionKernel.__call__
    return self._call(
cupy/_core/_reduction.pyx:351: in cupy._core._reduction._AbstractReductionKernel._call
    cub_success = _cub_reduction._try_to_call_cub_reduction(
cupy/_core/_cub_reduction.pyx:700: in cupy._core._cub_reduction._try_to_call_cub_reduction
    _launch_cub(
cupy/_core/_cub_reduction.pyx:536: in cupy._core._cub_reduction._launch_cub
    _cub_two_pass_launch(
cupy/_core/_cub_reduction.pyx:471: in cupy._core._cub_reduction._cub_two_pass_launch
    func = _SimpleCubReductionKernel_get_cached_function(
cupy/_util.pyx:67: in cupy._util.memoize.decorator.ret
    result = f(*args, **kwargs)
cupy/_core/_cub_reduction.pyx:243: in cupy._core._cub_reduction._SimpleCubReductionKernel_get_cached_function
    return _create_cub_reduction_function(
cupy/_core/_cub_reduction.pyx:228: in cupy._core._cub_reduction._create_cub_reduction_function
    module = compile_with_cache(
cupy/_core/core.pyx:2229: in cupy._core.core.compile_with_cache
    return cuda.compiler._compile_module_with_cache(
cupy/cuda/compiler.py:492: in _compile_module_with_cache
    return _compile_with_cache_cuda(
cupy/cuda/compiler.py:570: in _compile_with_cache_cuda
    ptx, mapping = compile_using_nvrtc(
cupy/cuda/compiler.py:321: in compile_using_nvrtc
    return _compile(source, options, cu_path,
cupy/cuda/compiler.py:290: in _compile
    options, headers, include_names = _jitify_prep(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
 
source = '/tmp/tmpuzshl335/2f864f67273a7ba84b6e0cac145efadc6ae13527.cubin.cu\n#include <cupy/complex.cuh>\n#include <cupy/carra...ge\n  }\n\n  if (_tid == 0) {\n      type_mid_out& out0 = *(_out0 + blockIdx.x);\n      POST_MAP(aggregate);\n  }\n}\n'
options = ('-DFIRST_PASS=1', '--std=c++11', '-DCUPY_USE_JITIFY', '-I/home/khushi/Documents/opensource/cupy/cupy/_core/include', '-I/home/khushi/Documents/opensource/cupy/cupy/_core/include/cupy/_cuda/cuda-11', '-I/usr/local/cuda/include', ...)
cu_path = '/tmp/tmpuzshl335/2f864f67273a7ba84b6e0cac145efadc6ae13527.cubin.cu'
 
    def _jitify_prep(source, options, cu_path):
        # TODO(leofang): refactor this?
        global _jitify_header_source_map_populated
        if not _jitify_header_source_map_populated:
            from cupy._core import core
            _jitify_header_source_map = core._get_header_source_map()
            _jitify_header_source_map_populated = True
        else:
            # this is already cached at the C++ level, so don't pass in anything
            _jitify_header_source_map = None
    
        # jitify requires the 1st line to be the program name
        old_source = source
        source = cu_path + '\n' + source
    
        # Upon failure, in addition to throw an error Jitify also prints the log
        # to stdout. In principle we could intercept that by hijacking stdout's
        # file descriptor (tested locally), but the problem is pytest also does
        # the same thing internally, causing strange errors when running the tests.
        # As a result, we currently maintain Jitify's default behavior for easy
        # debugging, and wait for the upstream to address this issue
        # (NVIDIA/jitify#79).
    
        try:
            name, options, headers, include_names = jitify(
                source, options, _jitify_header_source_map)
        except Exception as e:  # C++ could throw all kinds of errors
            cex = CompileException(str(e), old_source, cu_path, options, 'jitify')
            dump = _get_bool_env_variable(
                'CUPY_DUMP_CUDA_SOURCE_ON_ERROR', False)
            if dump:
                cex.dump(sys.stderr)
>           raise JitifyException(str(cex))
E           AssertionError: Only cupy raises error
E           
E           Traceback (most recent call last):
E             File "/home/khushi/Documents/opensource/cupy/cupy/cuda/compiler.py", line 255, in _jitify_prep
E               name, options, headers, include_names = jitify(
E             File "cupy/cuda/jitify.pyx", line 59, in cupy.cuda.jitify.jitify
E               cpdef jitify(str code, tuple opt, dict cached_sources=None):
E             File "cupy/cuda/jitify.pyx", line 92, in cupy.cuda.jitify.jitify
E               load_program(cuda_source, headers, nullptr, &include_paths,
E           RuntimeError: Runtime compilation failed
E           
E           During handling of the above exception, another exception occurred:
E           
E           Traceback (most recent call last):
E             File "/home/khushi/Documents/opensource/cupy/cupy/testing/_loops.py", line 45, in _call_func
E               result = impl(*args, **kw)
E             File "/home/khushi/Documents/opensource/cupy/tests/cupyx_tests/scipy_tests/special_tests/test_log_softmax.py", line 23, in test_log_softmax_large_inputs
E               return scp.special.log_softmax(a)
E             File "/home/khushi/Documents/opensource/cupy/cupyx/scipy/special/_logsumexp.py", line 71, in log_softmax
E               out = _log_softmax_kernel(tmp, axis=axis, keepdims=True)
E             File "cupy/_core/_reduction.pyx", line 752, in cupy._core._reduction.ReductionKernel.__call__
E               return self._call(
E             File "cupy/_core/_reduction.pyx", line 351, in cupy._core._reduction._AbstractReductionKernel._call
E               cub_success = _cub_reduction._try_to_call_cub_reduction(
E             File "cupy/_core/_cub_reduction.pyx", line 700, in cupy._core._cub_reduction._try_to_call_cub_reduction
E               _launch_cub(
E             File "cupy/_core/_cub_reduction.pyx", line 536, in cupy._core._cub_reduction._launch_cub
E               _cub_two_pass_launch(
E             File "cupy/_core/_cub_reduction.pyx", line 471, in cupy._core._cub_reduction._cub_two_pass_launch
E               func = _SimpleCubReductionKernel_get_cached_function(
E             File "cupy/_util.pyx", line 67, in cupy._util.memoize.decorator.ret
E               result = f(*args, **kwargs)
E             File "cupy/_core/_cub_reduction.pyx", line 243, in cupy._core._cub_reduction._SimpleCubReductionKernel_get_cached_function
E               return _create_cub_reduction_function(
E             File "cupy/_core/_cub_reduction.pyx", line 228, in cupy._core._cub_reduction._create_cub_reduction_function
E               module = compile_with_cache(
E             File "cupy/_core/core.pyx", line 2229, in cupy._core.core.compile_with_cache
E               return cuda.compiler._compile_module_with_cache(
E             File "/home/khushi/Documents/opensource/cupy/cupy/cuda/compiler.py", line 492, in _compile_module_with_cache
E               return _compile_with_cache_cuda(
E             File "/home/khushi/Documents/opensource/cupy/cupy/cuda/compiler.py", line 570, in _compile_with_cache_cuda
E               ptx, mapping = compile_using_nvrtc(
E             File "/home/khushi/Documents/opensource/cupy/cupy/cuda/compiler.py", line 321, in compile_using_nvrtc
E               return _compile(source, options, cu_path,
E             File "/home/khushi/Documents/opensource/cupy/cupy/cuda/compiler.py", line 290, in _compile
E               options, headers, include_names = _jitify_prep(
E             File "/home/khushi/Documents/opensource/cupy/cupy/cuda/compiler.py", line 263, in _jitify_prep
E               raise JitifyException(str(cex))
E           cupy.cuda.compiler.JitifyException: Runtime compilation failed
 
cupy/cuda/compiler.py:263: AssertionError
--------------------------------------- Captured stdout call ----------------------------------------
---------------------------------------------------
--- JIT compile log for /tmp/tmpuzshl335/2f864f67273a7ba84b6e0cac145efadc6ae13527.cubin.cu ---
---------------------------------------------------
cupy/complex/complex.h(94): warning: __device__ annotation is ignored on a function("complex") that is explicitly defaulted on its first declaration
 
cupy/complex/complex.h(101): warning: __device__ annotation is ignored on a function("complex") that is explicitly defaulted on its first declaration
 
specializations/block_reduce_raking_commutative_only.cuh(98): warning: declaration does not declare anything
 
/tmp/tmpuzshl335/2f864f67273a7ba84b6e0cac145efadc6ae13527.cubin.cu(134): error: more than one instance of overloaded function "exp" matches the argument list:
            function "exp(double)"
__nv_nvrtc_builtin_header.h(149892): here
            function "exp(float)"
__nv_nvrtc_builtin_header.h(150793): here
            argument types are: (const type_mid_in)
 
/tmp/tmpuzshl335/2f864f67273a7ba84b6e0cac145efadc6ae13527.cubin.cu(127): warning: variable "_J" was declared but never referenced
 
1 error detected in the compilation of "/tmp/tmpuzshl335/2f864f67273a7ba84b6e0cac145efadc6ae13527.cubin.cu".
 
---------------------------------------------------
dtype is <class 'numpy.int8'>
```

</p>
</details>

Reply:
It's because T is integer type (`char` from numpy.int8) here, so the compiler couldn't find `exp(char)`. Would you begin to try suppling float32 and float64 values? `@for_dtypes('fd')` in test instead of `@for_all_dtypes()`. Also related to https://github.com/cupy/cupy/pull/6823#issuecomment-1170645216 .

Reply:
Updated! Thanks :)